hpx/parallel/datapar/loop.hpp

```diff
@@ -0,0 +1,615 @@
+    {
+        return detail::datapar_loop<Begin>::call(begin, end, std::forward<F>(f));
+    }
+
+//     template <typename Begin, typename End, typename CancelToken, typename F>
+//     HPX_HOST_DEVICE HPX_FORCEINLINE Begin
+//     loop(parallel::v1::datapar_execution_policy, Begin begin, End end,
+//         CancelToken& tok, F && f)
+//     {
+//         return detail::datapar_loop<Begin>::call(begin, end, tok,
+//             std::forward<F>(f));
+//     }
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename ExPolicy, typename Begin, typename End, typename F>
+    HPX_HOST_DEVICE HPX_FORCEINLINE Begin
+    loop(ExPolicy&&, Begin begin, End end, F && f);
+
+    namespace detail
+    {
+        ///////////////////////////////////////////////////////////////////////
+        // Helper class to repeatedly call a function starting from a given
+        // iterator position.
+        template <typename Iter1, typename Iter2>
+        struct datapar_loop2
+        {
+            ///////////////////////////////////////////////////////////////////
+            template <typename InIter1, typename InIter2, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::enable_if<
+                iterators_datapar_compatible<InIter1, InIter2>::value &&
+                    iterator_datapar_compatible<InIter1>::value &&
+                    iterator_datapar_compatible<InIter2>::value,
+                std::pair<InIter1, InIter2>
+            >::type
+            call(InIter1 it1, InIter1 last1, InIter2 it2, F && f)
+            {
+                typedef typename hpx::util::decay<InIter1>::type iterator_type;
+                typedef typename std::iterator_traits<iterator_type>::value_type
+                    value_type;
+                typedef Vc::Vector<value_type> V;
+
+                if (detail::data_alignment(it1) || detail::data_alignment(it2))
+                {
+                    return std::make_pair(std::move(it1), std::move(it2));
+                }
+
+                InIter1 const last1V = last1 - (V::Size + 1);
+                while (it1 < last1V)
+                {
+                    datapar_loop_step2<InIter1, InIter2>::callv(f, it1, it2);
+                }
+
+                return std::make_pair(std::move(it1), std::move(it2));
+            }
+
+            template <typename InIter1, typename InIter2, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            static typename std::enable_if<
+                !iterators_datapar_compatible<InIter1, InIter2>::value ||
+                    !iterator_datapar_compatible<InIter1>::value ||
+                    !iterator_datapar_compatible<InIter2>::value,
+                std::pair<InIter1, InIter2>
+            >::type
+            call(InIter1 it1, InIter1 last1, InIter2 it2, F && f)
+            {
+                return std::make_pair(std::move(it1), std::move(it2));
+            }
```

Comment:
This has also been fixed now. As it turns out we need both versions, one which handles only those elements which are vectorizable, and one which handles the rest. 

Wrt equally misaligned iterators: not sure if this is worth implementing. I'd go ahead without it for now, it could be added later on, if really needed.